Expansion of skycultures through spatial and temporal extent

mRaetz opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

### Description
<!--- Please include a summary of the change and which issue is fixed. -->
<!--- Please also include relevant motivation and context. -->
<!--- List any dependencies that are required for this change. -->

This extension adds spatial and temporal dimensions to sky cultures. These are used to allow users to make interactive selections. The selection menu has been redesigned for this purpose. In addition to a list, users can now explore and select the available cultures using a time-resolved map. 
All new cultures in the future should have both a temporal and spatial dimension, which is why the SkyCultureMaker plugin has been updated as well. In addition to a region selection feature, a simple digitization tool has been implemented that allows users to digitize the culture territories themselves. Furthermore, a few minor changes were made that could be described as bug fixes. For example, a problem with pressing the ESC key in ScmSkyCultureDialog was fixed, as well as the previously invisible button for the artwork tooltips in ScmConstellationDialog.

- addition of spatial of temporal dimension to skyculture def
- new menu for skyculture selection that allows exploration of the available cultures
- integration of spatial / temporal extension into the SkyCultureMaker plugin
- small changes / bugfixes to the SkyCultureMaker plugin


### Type of change
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] This change requires a documentation update
- [ ] Housekeeping

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

This extension has been manually tested on a desktop PC (Windows 10) and a laptop (Windows 11). In both cases, there were no compatibility issues. However, no statement can be made regarding correct functionality on other operating systems.

Test specification:
- OS: Windows 10 / 11
- CPU: AMD Ryzen 7 2700x
- GPU: NVIDIA GeForce GTX 1660 Ti

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the [code style](http://stellarium.org/doc/head/codingStyle.html) of this project.
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (header file)
- [ ] I have updated the respective chapter in the Stellarium User Guide
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

mRaetz committed:
added custom GrahpicsView to display Map

added subclass of QGraphicsView to display a basemap (Pixmap) and several Polygons (GraphicItems)

mRaetz committed:
added custom GraphicsPolygonItem class

mRaetz committed:
registered new class files

mRaetz committed:
added Structure for new MapView

added QStackedWidget with 2 pages at the old location of SkyCultureTextBrowser. First page holds custom GraphicsView and a Slider, second page holds skyCultureTextBrowser.

mRaetz committed:
Merge branch 'master' into new-SkycultureMenu

mRaetz committed:
Merge branch 'master' into new-SkycultureMenu

mRaetz committed:
added Button to switch between SC map and browser

mRaetz committed:
fixed indentation

mRaetz committed:
skyCulture is now updated on selection of polygon in map

mRaetz committed:
time in skyculture map can now be changed

added QSlider and QSpinBox for time selection. Culture polygons are shown / hidden depending on the current year.

mRaetz committed:
skyCulture Polygon can be selected through list

When clicking on the culturesListWidget, a fitting polygon is selected in the SkycultureMapGraphicsView. If needed the current year is updated as well.

mRaetz committed:
Merge branch 'master' into new-SkycultureMenu

mRaetz committed:
changed baseMap picture to blank svg map

mRaetz committed:
Merge branch 'master' into new-SkycultureMenu

mRaetz committed:
added first attempt to load more complex polygon

loaded skyculture polygon from geojson file is visible on the map but not placed at the right position

mRaetz committed:
added transformation for projection

EPSG 3857 coordinates (meter) can now be transformed into view coordinates and projected onto the base map.
EPSG 4623 coordinates (lat / lon) can now be transformed into EPSG 3857 coordinates.

mRaetz committed:
added new base maps

added new SVG files (normal, compressed SVG and compressed into SVGZ format)

mRaetz committed:
removed map image for testing

mRaetz committed:
added smooth zoom

When a culture is selected through the listWidget, the view smoothly zooms out to the extent of the base map and then smoothly zooms onto the desired culture poygon

mRaetz committed:
adjustments to projection transformation

mRaetz committed:
changed some skyculture regions to be consistent with the UN geoscheme

mRaetz committed:
skyculture list is now sorted by UN geoscheme

mRaetz committed:
changed how sky culture description is accessed

removed switch page button and the stackedWidget. Added new page to ViewDialog with skyculture description.

mRaetz committed:
added search bar to skyculture list widget

mRaetz committed:
added Label for time

Added a QLabel that displays 'time' in the users respective language to the skyculture map page in ViewDialog

mRaetz committed:
enhanced the skyculture searchbar design

mRaetz committed:
added auto time init

Range of time for skyculture map (and Slider / SpinBox) is now initialized properly.

(at least the maxYear part which corresponds to the current system time)

mRaetz committed:
small convenience fixes

mRaetz committed:
changed zoom duration to scale with devation from default

mRaetz committed:
test GUI design for changeable time range

mRaetz committed:
added new Icons for SpinBox

added custom plus / minus icons for SpinBox up / down buttons

mRaetz committed:
changed the size constraints for culturesListWidget

mRaetz committed:
added new icons

mRaetz committed:
added startTime and endTime to StelSkyCulture class

mRaetz committed:
changed the skyCulture options GUI

Moved the skyCulture (Constellation) Options to the skycultureDescription page. Added new time / culture related Options to the skycultureMap page.

mRaetz committed:
added time initialization function

The newly added initSkycultureTime function sets all time related components properties.

(Note: minTime is not set correctly, but to an arbitrary test value)

mRaetz committed:
added dynamically changeable time limits

minTime- and maxTimeSpinBox can now set the limits of timeSlider and currentTimeSpinBox.

(And if necessary change the current time value)

mRaetz committed:
time limits can now be used to filter the culture list

mRaetz committed:
Merge branch 'master' into new-SkycultureMenu

mRaetz committed:
UI improvements for the Skyculture (time) options

mRaetz committed:
UI improvement: added CheckBox to allow for map rotation

mRaetz committed:
changed how polygon item selection behaves

SkyculturePolygonItems now stay selected when visibility is set to false

mRaetz committed:
fixed map zoom bug on dialog creation

When the contents of the ViewDialog are created a bug occured which resulted in a very small initial zoom level. This happened because the scale ratio was calculated using the current size of the Widget before it was properly created.

mRaetz committed:
on selection: select all polygons of a culture

The selection of any culture polygon now properly selects all respective polygons belonging to this culture (at different timestamps)

mRaetz committed:
updated all previous made connections to modern function pointers

mRaetz committed:
reimplemented GraphicsView's DragMode

reimplemented / customized mouse: move / press / release event for better controls when moving the map

mRaetz committed:
added smaller zoom when holding CTRL

mRaetz committed:
added map rotation and removed some leftovers

mRaetz committed:
set current culture time limit labels when updating the default SC

mRaetz committed:
Merge branch 'master' into new-SkycultureMenu

mRaetz committed:
Merge branch 'master' into new-SkycultureMenu

mRaetz committed:
removed old test files

mRaetz committed:
added ComboBox component that supports selection of multiple items

mRaetz committed:
added preview items for ScMaker map (polygon digitizing)

mRaetz committed:
added custom Tree item for polygon info in ScMaker

mRaetz committed:
added new Region type for combobox in ScMaker

mRaetz committed:
one culture can have multiple regions

Instead of only one region, every culture can have multiple regions. This new format is stored as a JsonArray but the legacy version with JsonObject is still supported.

mRaetz committed:
start and end time labels in ScMenu are correctly set

mRaetz committed:
updated skyculturePolygonItem visual appearance

mRaetz committed:
load all available culture polygons from the respective json files

mRaetz committed:
added new CulturePolygon type for ScMaker

mRaetz committed:
added geoLoc tab to ScMaker

Added GeoLoc Tab to ScMaker Plugin. This allows the user to digitize and save culture polygons.

mRaetz committed:
small changes pre merge

changed some debug messages, names, comments, ...

mRaetz committed:
Merge branch 'master' into new-SkycultureMenu

mRaetz committed:
renamed files to match other SCMaker file names

mRaetz committed:
added final map image for SCE / SCM (+ removed test file)

mRaetz committed:
added new custom SpinBox class for SCMaker

mRaetz committed:
revised tooltip in SCMaker artwork dialog

The button was invisible and the text was displayed as a tooltip.

- moved the button to the top right corner of the artwork editior
- changed the icon (now visible)
- replaced the tooltip approach with a label (actually a QPushButton) that will be shown when the button is clicked

mRaetz committed:
added some comments and removed some debug messages

mRaetz committed:
changed the tooltip for the map in SCMaker from QToolTip to label

mRaetz committed:
changed name of some components in ui file

mRaetz committed:
SCMaker / SCExplorer now support existing cultures

Previously, existing cultures could not be used because the endTime was stored as a number (what happens when a new year begins?). Now, the endTime is stored as a string so that this special case can be encoded with the infinity symbol (∞).

mRaetz committed:
resolved bug when using ESC key instead of titlebar

Installed an event filter so ESC key doesn't reject the dialog anymore

mRaetz committed:
added new controls for map navigation and new point

mRaetz committed:
added smooth zoom for multiple culture polygons

If several polygons of a culture exist at the same time, the zoom will no longer focus on just one of these polygons, but on the bounding box around them.

mRaetz committed:
changed (string) filter in search bar for cultureListWidget

Implemented modified damerau-levenshtein distance to correct minor typos made by users.
(Not perfect but works: in some cases it doesn't hide totally different words because insertions are not penalized at all.)

mRaetz committed:
cleanup (remove debug messages, comments, ...)

mRaetz committed:
Merge branch 'master' into new-SkycultureMenu

mRaetz committed:
Merge branch 'master' into new-SkycultureMenu

mRaetz committed:
added / removed some comments

alex-w commented:
Oops... @mRaetz please resolve conflicts

mRaetz committed:
changed removal of locations to safe pattern

The previously used procedure could lead to undefined behavior under certain circumstances.

mRaetz committed:
Merge branch 'master' into new-SkycultureMenu

mRaetz committed:
remove old eventFilter implementation

A new evenFilter was installed that fixes the same problem (b10f9c4)

xLPMG commented:
I compiled this on my machine and your code generates a few new warnings that need to be fixed. Some examples include

```
ScmSkyCultureDialog.cpp:659:69: warning: unused parameter 'column' [-Wunused-parameter]
  659 | void ScmSkyCultureDialog::selectLocation(QTreeWidgetItem *item, int column)
      |                                                                     ^
ScmMultiselectionComboBox.cpp:131:12: warning: variable 'currentModel' set but not used [-Wunused-but-set-variable]
  131 |         if (auto *currentModel = model)
      |                   ^
SkycultureMapGraphicsView.cpp:452:8: warning: unused variable 'maxDuration' [-Wunused-variable]
  452 |         qreal maxDuration = 2000;
      |               ^~~~~~~~~~~
SkyculturePolygonItem.cpp:30:4: warning: field 'isHovered' will be initialized after field 'lastSelectedState' [-Wreorder-ctor]
   30 |         , isHovered(false)
      |           ^~~~~~~~~~~~~~~~
      |           lastSelectedState(false)
   31 |         , lastSelectedState(false)
      |           ~~~~~~~~~~~~~~~~~~~~~~~~
      |           isHovered(false)
```

If these dont appear in your console, maybe try deleting the build folder and doing a fresh build.

Also now that I see this warning: the usual name scheme was to use camel case for "SkyCulture", but your files use "Skyculture". Its only a cosmetic issue, but it would still be nice if the same naming scheme is used everywhere :))

mRaetz committed:
corrected naming of classes and fixed some warnings

mRaetz commented:
Thanks for the tip and for pointing out the incorrect names. I didn't even notice the mistake with the lowercase c.

Another thing i noticed when looking over the files is that one of the buttons in ScmHideOrAbortMakerDialog is incorrectly translated in german. The buttons are supposed to be distinguishable (abort - hide - cancel) but in the current version they are not (Ausblenden - Abbrechen - Abbrechen)
 
<img width="712" height="176" alt="{9A74251C-C552-4516-9271-54471F9B0311}" src="https://github.com/user-attachments/assets/3cea61f5-73bc-4c0e-82a3-e2709a271fc3" />

I'm also not sure how to solve the problem with the SvgWidgets module. Apparently, there was a change in Qt in this regard, but even if there is a difference between Qt5 and Qt6, I don't understand why the checks for Linux (amd64; qt6; core) / Linux (amd64; qt6) failed when they passed for Windows and macOS with qt6.

xLPMG committed:
test ci

xLPMG committed:
change file case

xLPMG committed:
try some fixes

xLPMG committed:
conditional svgwidgets

xLPMG committed:
install different package

xLPMG committed:
Update ci.yml

xLPMG committed:
linux qt6

xLPMG committed:
fix qt5 errors

xLPMG committed:
more qt5 issues

xLPMG committed:
fix other ci steps

xLPMG committed:
fix more issues

xLPMG committed:
add QMath

xLPMG committed:
test

xLPMG committed:
test #2

mRaetz committed:
Merge pull request #27 from xLPMG/qt-svg-widgets

fixed some build issues and warnings

mRaetz committed:
fixed some warnings regarding compability in qt5

- fixed error in Qt5 where QList<QPointF> cannot be converted into a QPolygonF
- previous change of class name left behind artifact in ui file
- changed a small visual problem of the preview when setting the first point of a polygon (in SkyCultureMaker)

10110111 commented on plugins/SkyCultureMaker/src/gui/ScmGeoLocGraphicsView.cpp:
```diff
@@ -0,0 +1,386 @@
+#include <QJsonObject>
+#include <QJsonDocument>
+#include <QtMath>
+#include <cmath> // for M_PI
```
`<QtMath>` is documented to define `M_PI` too. But why on Earth use `QtMath` at all? For `qPow`? How is it better than `std::pow`?

Besides the apparent uselessness, Qt is constantly making some crazy assumptions, like `qFloor` being templated (so that it could support large numbers) but returning a mere `int` (throwing away most of supportable integers), downcast from a call to `std::floor` (basically)!

So I'd advise to avoid `QtMath`, except maybe for the unique functions like the variadic `qHypot` overload or `qNextPowerOfTwo`, etc.

10110111 commented:
I had to apply the following patch to build without precompiled headers on Ubuntu 20.04 and Qt 6.4.1:
```diff
diff --git a/src/gui/SkyCultureMapGraphicsView.cpp b/src/gui/SkyCultureMapGraphicsView.cpp
index c5fcee207a..edaa5faaec 100644
--- a/src/gui/SkyCultureMapGraphicsView.cpp
+++ b/src/gui/SkyCultureMapGraphicsView.cpp
@@ -20,6 +20,7 @@
 #include "SkyCultureMapGraphicsView.hpp"
 #include "SkyCulturePolygonItem.hpp"
 #include "StelLocaleMgr.hpp"
+#include "StelFileMgr.hpp"
 #include "StelSkyCultureMgr.hpp"
 #include <qjsonarray.h>
 #include <QGraphicsSvgItem>
@@ -27,7 +28,10 @@
 
 #include <QJsonObject>
 #include <QJsonDocument>
+#include <QWheelEvent>
+#include <QMouseEvent>
 #include <QtMath>
+#include <QFile>
 
 SkyCultureMapGraphicsView::SkyCultureMapGraphicsView(QWidget *parent)
     : QGraphicsView(parent)
```

10110111 commented:
So, I've finally tried this PR. Here's my critique:

0. See the above comment.
1. The separate "Sky Culture Description" looks quite out of place as a category among the other tabs. Would it be more appropriate to make it a subtab inside the "Sky culture" tab?
2. The labels like "current Time", "min Year" etc. should be capitalized in a way to match the rest of the GUI, i.e. "Current time", "Minimum year" etc. (note "minimum" instead of "min", because there's no reason to shorten it, since we have plenty of space in this row).
3. Why does the "current Time" spinbox look so different from all the other spinboxes? The -/+ buttons seem to do exactly the same as the usual down/up buttons.
4. The "Enter Culture..." filter behaves strangely: I type `b`, it shows all SCs that contain a letter "b" in their names, but if I type `be`, almost all SCs reappear (but some are still filtered out).
5. I can move the world map by half its size up and down, showing useless space above or below, which makes no sense to me.
6. Sky Culture Description contains most of the settings, which is not obvious from tab name. This implies that the name should rather be something like "Sky Culture Details".

mRaetz commented:
> The separate "Sky Culture Description" looks quite out of place as a category among the other tabs. Would it be more appropriate to make it a subtab inside the "Sky culture" tab?

Originally, the map and description were in the same tab (a stackedWidget allowed users to switch between them). On the suggestion of GZotti / SMH, the description was later moved to this extra tab.

> Sky Culture Description contains most of the settings, which is not obvious from tab name. This implies that the name should rather be something like "Sky Culture Details".

Yes, the name could definitely be better. If the description stays in its own tab, I would also prefer this name.

> Why does the "current Time" spinbox look so different from all the other spinboxes? The -/+ buttons seem to do exactly the same as the usual down/up buttons.

This was purely a design decision. I felt that buttons on the left and right would better convey the connection with the horizontal slider. The normal up/down buttons rotated and placed on the left/right looked a bit strange, which is why I used -/+ buttons. If this disturbs the overall design too much, the normal up/down buttons on the right side of the spin box can be used instead.

> The "Enter Culture..." filter behaves strangely: I type b, it shows all SCs that contain a letter "b" in their names, but if I type be, almost all SCs reappear (but some are still filtered out).

Yes, that's right. I initially used a strict query (toLower(filterString) must be contained in the word). While testing, I noticed that some cultures are difficult to find with this query. Typos are an obvious problem, but special characters can also prevent the filter from matching the word. A concrete example of this is as follows: In the German version, the indigenous culture of New Zealand is referred to as Māori. However, the ā is a very rare character, so it must be assumed that the user will enter Maori (normal a) in the search bar. Therefore, I tried to use a kind of string similarity measure to solve this problem. This is not perfect, but above all, it does not filter individual characters (a deviation of at least one character should be allowed to avoid the problems described above). Therefore, it is only used for 2 or more characters, which leads to the behavior described. I am sure there is a better solution for the whole thing, but the current implementation has basically done its job. However, I would also be open to using a simple, strict matching (via contains).

> I can move the world map by half its size up and down, showing useless space above or below, which makes no sense to me.

If the scene is the same size as the map, navigation of the map becomes restricted. On the one hand, you could then only view the edge of the map (depending on the zoom level) at the edge of the window, but most of the time you want the relevant part to be in the center of your field of view. On the other hand, there is a problem with smooth zooming after selecting a culture from the list. Since during this zoom only the scene is moved to create the illusion of smooth movement, it occurs, especially in the border areas of the map, that you reach the boundary of the scene, resulting in a rigid movement. This can be solved by setting the extent of the scene to a multiple of the map.

mRaetz committed:
Merge branch 'master' into qt5-compability-fixes

mRaetz committed:
fix appearance to fit the style of stellarium

- reverted custom SpinBox design
- created sub-tab for skyculture details / map page
- changed field names to begin with upper case letters

mRaetz committed:
fixed search bar implementation

changed search bar to simply ignore diacritics instead of similarity estimation with string measures

mRaetz committed:
restrict map movement

the map now cant be moved out of the view (locks at half of the screen)

mRaetz committed:
cleanup digitized polygons before exporting

To prevent potential user error all polygons are checked before the culture is exported. If necessary, overlapping polygons are merged to and adjusted.

mRaetz committed:
fix some comments

mRaetz committed:
fixed compatibility issues with other qt versions

mRaetz committed:
basemap now features some lakes

mRaetz committed:
Merge branch 'master' into qt5-compability-fixes

mRaetz committed:
replace qMath functions with cmath

mRaetz committed:
more qMath replacement that was overlooked

gzotti commented on skycultures/lokono/index.json:
```diff
@@ -1,6 +1,6 @@
 {
   "id": "lokono",
-  "region": "Southern America",
+  "region": "South America",
```
What is the reason to deviate from the list documented and published before in the User Guide, section 9.2.1?

gzotti replied:
OK I see it. Not sure who introduced the list in its current form, but now we have used it for a year already and IMHO it should not be changed unless really wrong (like "Suothern Amreica"). Same goes for "World" vs "Global". There are several projects already utilizing tens of skycultures in the form documented in the User Guide. Other opinions, @10110111 , @alex-w, (probably) last chance to change, or keep using the current 2025 spelling?

10110111 replied:
I think backwards compatibility is better than pursuit of all the changes in wording upstream (although "Southern America" does sound odd).

alex-w replied:
It may be my orthographic mistake (South America) due I filled the list of regions - I'll fix it in the master.

gzotti commented on src/gui/ViewDialog.hpp:
```diff
@@ -139,6 +146,14 @@ private slots:
+
+	// list of special characters and their 'normal' counterpart for filtering (skycultures / the culturesListWidget)ĀA
+	// (when a new special character is added to 'specialCharString', 'normalCharList' must also be expanded)
+	QString specialCharString = "ŠŒŽšœžŸ¥µÀÁÂÃĀÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖØÙÚÛÜÝßàáâãāäåæçèéêëìíîïðñòóôõöøùúûüýÿ";
```
There is at least a "dotless i" to be included. But see Vietnamese for a few more suggestions :-)

mRaetz replied:
I'm not entirely sure anymore because I didn't write down the results back then, but I experimented with the normalization methods before the current implementation, and none of them delivered a completely satisfactory result. (I believe it was mostly about the ligatures)
In any case, it will probably be difficult to cover all special characters in this way. I'm not quite sure how to handle this yet, whether all existing special characters should appear in the list as a preventive measure, or whether it should be expanded in the future if problems arise. (as in minor problems because it is a QoL feature that won't cause any issues apart from inconveniences)

gzotti commented:
In the SCM GUI, I see a few weird things.

- you can select a classification. It seems you did not write this code, but IMHO we should make sure that NONE is not an option here. A newly created SC should be personal for a start, and authors can select something else where applicable. 
- You can apparently select more than one region. The region entry in index.json only has one string. Huh?

The Constellation Editor/Information is missing a line for the (optional, scientific) Transliteration (see User Guide 9.2.4). 

The polygon drawing works well but still could become more intuitive. Of course it depends on which other software you are using to use the same keys.  I had expected LMB draws, and RMB finalized one polygon, and needed the help button.  You could remove the Alt combination, and use ESC as panic button to remove last-pressed key. Remove full polygon is available in the dedicated button, so may also not be necessary on a key. Or compare to the existing constellation drawing in the same plugin: RMB to draw, double click to finish. This could be harmonized.

Who is going to write the detailed chapter for the User Guide? The logic behind the temporal polygons needs explanation.

gzotti commented:
> 1. Classification Combobox:
> 
> [...] If the user doesn’t [select a valid option], a pop-up window will appear when exporting Skyculture, informing them that a classification != none must be selected.

OK, seen. 
 
> 2. Multi-Region Skycultures:
> 
> > You can apparently select more than one region. The region entry in index.json only has one string. Huh?
> 
> In my discussions with @sushoff , [...], all cultures should have the new region format.

What about the other projects utilizing the new SC format? @10110111 is S-M able to cope with this? The only format relevant as of 25.4 is the one from the User Guide.

 
> 3. Constellation Editor missing field
> 
> > The Constellation Editor/Information is missing a line for the (optional, scientific) Transliteration (see User Guide 9.2.4).
> 
> This appears to be an error on the part of the original SCM team. I have no idea whether this field was simply overlooked or whether there was a reason not to implement it.

Must have been overlooked, and should be added.

> 4. Controls of polygon digitizing
> 
> 
> I tried to base the controls on those in QGIS, simply because they are already established there. Originally, a new point could only be set with LMB, but based on feedback that the Constellation Editor uses RMB, I also allowed RMB to be used to set a new point. Personally, I prefer the option where you add a point with LMB and finalize the polygon with RMB. (I'm not a fan of the controls in the Constellation Editor, but my personal preferences shouldn't matter here. If there is a desire for uniform controls, I will be happy to implement that.) [...]

For me it appears desirable to have similar controls for similar actions in a single plugin. @alex-w ? @10110111 ?

> It's true that a polygon can be deleted once it has been digitized, but I don't really see why it would be necessary to remove the option to reset the digitization process (in the end its just an optional QoL feature). Again, this is just my personal opinion, because that's how I've come to use it in other programs (such as QGIS or ArcGIS), but I think using the backspace key to delete the last point is more intuitive than using ESC.

Sure, ESC is uncommon. Backspace is OK for me.

 
> 5. User Guide
> 
> > Who is going to write the detailed chapter for the User Guide? The logic behind the temporal polygons needs explanation.
> 
> I suppose I have to at least help the person writing the chapter in the form of a detailed explanation (if not more).

I can do some formatting, but the operational instructions should come from someone who has designed this or who wants other people to use it, else nobody will know what to do with it. @sushoff ? 
 
> > 6. Unfortunately, the program then exited silently. Also after relaunch, calling up the view menu to load the new SC failed with silent exit.
> 
> I have never encountered the bug [...]

For now my creation seems it work at least after restart. But before merge the plugin should survive storing SC and reloading all SCs, then loading and showing the newly created one.

mRaetz committed:
add missing "transliteration" field in SCM

mRaetz committed:
change controls for poly digitizing in SCM

mRaetz committed:
revert name change of region

mRaetz committed:
expand special character list for sc filter in ViewDialog

mRaetz committed:
update sc tab in ViewDialog after export in SCM

mRaetz committed:
revert multiple region export

Left possibility to use multiple regions for one skyculture in the code. The region is saved as a string again (instead of an array).

mRaetz committed:
add alternative control to delete the last added point

mRaetz committed:
fix bug in new controls in SCM

mRaetz committed:
some function refactoring and cleanup

mRaetz commented:
In response to @gzotti in #4636:

The missing territory files are actually optional; thanks for pointing that out. I’ve changed it to qWarning. 
Since the program crashed again for you: During development, I only used the skycultures already available in Stellarium for testing. There don't seem to be any issues with those. Could you send me your collection of SCs (or at least a part if it) so I can investigate the source of the error? (Assuming this collection of yours differs from the skycultures on the main branch.)

gzotti commented:
The other question was, who will add all the polygons for the existing SCs so that the map selection becomes the awesome feature it now only promises to become? 
While we are changing, when viewing a territory.json I see the keys are stored alphabetically, as usual. However,  when wanting to edit this file in a text editor, this gives an eye-discomforting sectioning per polygon: endTime, geometry, startTime. If you could rename startTime to beginTime, it would be placed into logical order. 

Another issue is when adding a polygon in March 2026.  endTime=2026. What is the interpretation if I look into this in 2030? Is it useful to add special values like "origins" (as hypothetical unknown before date) and  "present", or is it acceptable to just leave endtime away to interpret as "now"?

On the technical side: There is the GeoJSON fileformat. Wouldn't it be useful to follow this (expandable) simple JSON-based standard, that might make editing with other tools easier? (Imagine editing a detailed shapefile over a detailed map in QGIS and exporting in GeoJSON. We could have a "master shapefile" with per-shape fields "skyculture", "beginDate", "endDate" in which we collect all world entries.) Each current polygon would be a feature in this FeatureCollection, with a Geometry of type polygon, with properties beginTime and EndTime. This (utilizing GeoJSON) would be a change I'd like to see for 26.2 (while maps are not yet really useful in 26.1) when we can accept the rest now. The QGIS map is something for the future (2026-2027), as it should come with its own import/export QGIS plugin.

mRaetz committed:
changed territory file output to warning since its optional

mRaetz committed:
accept "old" region key of South America

mRaetz committed:
changed name of var and fixed bug

- changed the name of `startTime` to `beginTime` for better readability in the JSON files (beginTime --> endTime --> polygon)
- fixed a bug when deleting a polygon through the delete button in SCM (sometimes the wrong endTime was displayed)
- changed the logic of the addPolygonDialog in SCM (when should a polygon have no fixed endTime) from > current year to >= current year

mRaetz committed:
Merge branch 'master' into new-SkycultureMenu

mRaetz committed:
added support for geojson territory files in SCM

Added functionality to write GeoJSON files instead of regular JSON files when exporting the territory of the skyculture.

gzotti commented:
> > I have a preference to use existing applicable standard formats and switch before the old format is being circulated
> 
> That sounds to me like we're going to use GeoJSON directly? If so, I'll push the changes so that the territory.geojson files are read correctly. (?)

If nobody objects by tomorrow (@alex-w , @10110111 , @sushoff ?), I see no point in keeping the self-made format. 

> > I have just pushed an editor-converted territory.geojson of norse_edda to master.
> 
> It's looking pretty good, but im afraid those weren't the only issues. The endTime is still an int (while the current implementation uses a string), 

Ah, you can change the property type then. As I understand the properties are a dictionary where the data types are defined as required. However, you could also just use 10.000 as special marker value. 

> and the entries for the global beginTime and endTime are missing from the index.json file.

Where should it go? It is not described in the specifications (User Guide), so I know nothing about it. 
 
> Another thing (that isn't really due to the format, but it's probably something that could have been avoided): There's an artifact between the polygons that likely has no business being there, but is instead a result of the digitization process. 

Right, but it was already there. Maybe we can repair this into a 4th polygon feature later. I was not aware of this file until 2 hours ago...

mRaetz committed:
switch territory file format from plain JSON to geoJSON

mRaetz committed:
fixed format / artifact in index and territory file

mRaetz committed:
change endTime from string back to int

Switch back from using a string for endTime to using an int again. The new special value for existing skycultures is 9146 (previously it was "∞").

mRaetz committed:
Merge branch 'master' into new-SkycultureMenu

gzotti commented:
Map: Each SC can now have a GeoJSON file territory.geojson that shows its location as click-selectable polygon. The time slider below would show the localisation of the SCs as coded in beginTime/endTime properties. The polygon can be created when defining a SC with the SCM plugin, and probably edited/refined with general GIS tools like QGIS.  So far, only Norse-Edda has it, but we should add those files to all non-"World" SCs in the upcoming months to populate the map (#4842). I have disabled a warning for missing polygon for now.

axd1967 commented:
@mRaetz Very nice but I never understood why project managers accept a change without corresponding user manual updates. 

There is even a checkbox... it should be checked...

It is part of the development effort that should not be left to others to fix. Whi else will do it? Who has the best knowledge of what changed? At least a few sentences should be added, even if it's only a "TODO", "update screenshot here", add some basic explanation...

I know by experience that developers tend to underestimate the value of the UG... They like to code, and leave user manual updates to others.

gzotti commented:
@axd1967 project managers have decided to accept this as-is for 26.1 and add notes into the SUG only when at least a few map polygons have been added so that it's worth discussing them. (Unfortunately I have not seen _someone_'s PR with ~35 GeoJSONs by now - maybe _anyone_ could do it instead...) But generally I agree, new features are best described by the respective developer which is why the SUG is a "living" part of the git and not a separate project as it was 15-10 years ago.

sushoff commented:
@axd1967 you know, unpaid project managers just want to punish the world ... 

@gzotti is right: we are currently working on adding data:  it's correct that there hasn't been any PR yet, but just a couple of hours ago, we discussed it (there had been some delay in the past ~2 weeks due to illness of humans who are unfortunately not able to work always at same pace in spare time activities).
